galley/pkg/config/analysis/analyzers/webhook/webhook.go

```diff
@@ -0,0 +1,214 @@
+		for _, base := range getObjectLabels() {
+			base[label.IoIstioRev.Name] = rev
+			objectLabels = append(objectLabels, base)
+		}
+	}
+
+	// For each permutation, we check which webhooks it matches. It must match exactly 0 or 1!
+	for _, nl := range namespaceLabels {
+		for _, ol := range objectLabels {
+			matches := sets.NewSet()
+			for name, whs := range webhooks {
+				for _, wh := range whs {
+					if selectorMatches(wh.NamespaceSelector, nl) && selectorMatches(wh.ObjectSelector, ol) {
+						matches.Insert(fmt.Sprintf("%v/%v", name, wh.Name))
+					}
+				}
+			}
+			if len(matches) > 1 {
+				for match := range matches {
+					others := matches.Difference(sets.NewSet(match))
+					context.Report(webhookCol, msg.NewInvalidWebhook(resources[match],
+						fmt.Sprintf("Webhook overlaps with others: %v. This may cause injection to occur twice.", others.UnsortedList())))
+				}
+			}
+		}
+	}
+
+	// Next, check service references
+	for name, whs := range webhooks {
+		for _, wh := range whs {
+			if wh.ClientConfig.Service == nil {
+				// it is an url, skip it
+				continue
+			}
+			fname := resource.NewFullName(
+				resource.Namespace(wh.ClientConfig.Service.Namespace),
+				resource.LocalName(wh.ClientConfig.Service.Name))
+			if !context.Exists(serviceCol, fname) {
+				context.Report(webhookCol, msg.NewInvalidWebhook(resources[fmt.Sprintf("%v/%v", name, wh.Name)],
+					fmt.Sprintf("Webhook refers to a service that does not exist: %v.", fname)))
+			}
+		}
+	}
+}
+
+func isIstioWebhook(wh *v1.MutatingWebhookConfiguration) bool {
+	for _, w := range wh.Webhooks {
+		if strings.HasSuffix(w.Name, "istio.io") {
+			return true
+		}
+	}
+	return false
+}
+
+func extractRevisions(wh *v1.MutatingWebhookConfiguration) []string {
+	revs := sets.NewSet()
+	if r, f := wh.Labels[label.IoIstioRev.Name]; f {
+		revs.Insert(r)
+	}
+	for _, webhook := range wh.Webhooks {
+		if webhook.NamespaceSelector != nil {
+			if r, f := webhook.NamespaceSelector.MatchLabels[label.IoIstioRev.Name]; f {
+				revs.Insert(r)
+			}
+		}
+		for _, ls := range webhook.NamespaceSelector.MatchExpressions {
+			if ls.Key == label.IoIstioRev.Name {
+				revs.Insert(ls.Values...)
+			}
+		}
+		if webhook.ObjectSelector != nil {
+			if r, f := webhook.ObjectSelector.MatchLabels[label.IoIstioRev.Name]; f {
+				revs.Insert(r)
+			}
+
+			for _, ls := range webhook.ObjectSelector.MatchExpressions {
+				if ls.Key == label.IoIstioRev.Name {
+					revs.Insert(ls.Values...)
+				}
+			}
+		}
+	}
+	return revs.UnsortedList()
+}
+
+func selectorMatches(selector *metav1.LabelSelector, labels klabels.Set) bool {
+	// From webhook spec: "Default to the empty LabelSelector, which matches everything."
+	if selector == nil {
+		return true
+	}
+	s, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return false
+	}
+	return s.Matches(labels)
+}
+
+//
+//func (a *Analyzer) analyzeProtocolAddresses(resource *resource.Instance, context analysis.Context) {
```

Comment:
can you remove this code if it is no longer needed?

Reply:
done